pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java

```diff
@@ -488,6 +488,10 @@ public void sendAsync(Message<?> message, SendCallback callback) {
             return;
         }
 
+        // Update the message metadata before computing the payload chunk size to avoid a large message cannot be split
+        // into chunks.
+        final long sequenceId = updateMessageMetadata(msgMetadata, uncompressedSize);
```

Comment:
There are still some metadata updates after this method, like `msgMetadata.setUuid(...)` and `msgMetadata.setChunkId().....` and even in `ProducerImpl#encryptMessage`.

Can we just skip `ProducerImpl#isMessageSizeExceeded` if client chunking is enabled.

Reply:
Yes, we can skip `isMessageSizeExceeded` if client chunking is enabled. But I think the existing changes in this PR make the computation of the payload chunk size **more accurate**.

Regarding to the metadata update after the computation,

>  msgMetadata.setUuid(...) and msgMetadata.setChunkId().....

These metadata fields are updated only for chunks:

```protobuf
    optional string uuid = 26;
    optional int32 num_chunks_from_msg = 27;
    optional int32 total_chunk_msg_size = 28;
    optional int32 chunk_id = 29;
```

In this case, `isMessageSizeExceeded` is already skipped.

And yes, `ProducerImpl#encryptMessage` could update the metadata as well. I think it's better to skip the message size check when encryption is enabled, because message encryption could also increase the size of the payload buffer.

Reply:
IMO, it's also better to make it clear which metadata fields can be set before checking if the size exceeds the limit.

Reply:
> But I think the existing changes in this PR make the computation of the payload chunk size **more accurate**.

+1

Reply:
> IMO, it's also better to make it clear which metadata fields can be set before checking if the size exceeds the limit.

I think the bottom line is the fields which can be very large must be included here, like `property`. Other fields won't affect much as they take constant small spaces.

Reply:
> `ProducerImpl#encryptMessage` could update the metadata as well. I think it's better to skip the message size check when encryption is enabled, because message encryption could also increase the size of the payload buffer.

This is a little out the scope of this PR. Maybe need more discussion. I am not sure if the space complexity of the increased size is O(1) compare to the limitation. If not, I think it's better to check the size at client side.